### PR TITLE
Adding dev-spaces commands under aks module in azure cli

### PR DESCRIPTION
**Pull request by saurabsa**
- use_dev_spaces
        - Calling out to install dev-spaces-preview extension and calling ads_use_dev_spaces
- remove_dev_spaces
        - Calling ads_remove_dev_spaces from dev-spaces-preview extension module
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/6369
This is an **experimental preview** for @microsoft users.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py [hunk 1]
```diff
@@ -384,6 +384,39 @@
               --connector-name aci-connector --chart-url <CustomURL> --image-tag <VirtualKubeletImageTag>
 """
 
+helps['aks use-dev-spaces'] = """
+    type: command
+    short-summary: (PREVIEW) Use Azure Dev Spaces with a managed Kubernetes cluster.
+    parameters:
+        - name: --name -n
```
Comment (stepro) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: I don't see any other place in this file where the -n, -g options are explicitly documented. I think they get inherited from some parent context. What happens if you remove them? The main reason I ask is that the short-summary values you provide are lacking newlines in appropriate places to make sure the lines don't get too long. I think by letting something else include these, the wrapping will be done properly.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py [hunk 2]
```diff
@@ -384,6 +384,39 @@
...
+          short-summary: Name of the managed cluster.
+        - name: --resource-group -g
+          type: string
+          short-summary: Name of resource group. You can configure the default group using 'az configure --defaults group=<name>'.
+        - name: --space -s
+          type: string
+          short-summary: Name of the dev space to use.
+        - name: --parent-space -p
+          type: string
+          short-summary: Name of a parent dev space to inherit from when creating a new dev space. By default, if there is already a single dev space with no parent, the new space inherits from this one.
```
Comment (stepro) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: This short-summary at least needs to manually implement wrapping. Looking at other help documentation in az, it seems they have standardized on wrapping at 100 characters per line.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py (hunk 1 above)
Comment (stepro) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: Look at _params.py. There's some "aks" context that defines the -g and -n options so you don't have to.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 3]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
     return sdk_no_wait(no_wait, client.create_or_update, resource_group_name, name, instance)
 
 
+def aks_use_dev_spaces(cmd, client, cluster_name, resource_group_name, space_name='default', parent_space_name=None):  # pylint: disable=line-too-long
+    """
+    Use Azure Dev Spaces with a managed Kubernetes cluster.
```
Comment (stepro) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Should this be prefixed with (PREVIEW)?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 4]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+    :param space_name: Name of the dev space to use.
+    :type space_name: String
+    :param parent_space_name: Name of a parent dev space to inherit from when creating a new dev space. \
+    By default, if there is already a single dev space with no parent, the new space inherits from this one.
+    :type parent_space_name: String
+    """
+
+    dev_spaces_extension = 'dev-spaces-preview'
+
+    should_create_resource = False
```
Comment (stepro) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Name is confusing and not related to the conditions you have below. Why does this condition exist? I see no case where "az aks use-dev-spaces" shouldn't delegate to the extension's use_dev_spaces function.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 5]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+    :param resource_group_name: Name of resource group. You can configure the default group. \
+    Using 'az configure --defaults group=<name>'.
+    :type resource_group_name: String
+    :param prompt: Do not prompt for confirmation.
+    :type prompt: bool
+    """
+
+    dev_spaces_extension = 'dev-spaces-preview'
+
+    should_delete_resource = False
```
Comment (stepro) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Same comment on this condition.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 4 above)
Comment (stepro) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: I see. You should call this variable something like "extension_installed".

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 6]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+            # Adding the installed extension in the path
+            from azure.cli.core.extension import get_extension_path
+            ext_dir = get_extension_path(dev_spaces_extension)
+            sys.path.append(ext_dir)
+            # Import the extension module
+            from importlib import import_module
+            azext_custom = import_module('azext_dev_spaces_preview.custom')
+            azext_custom.ads_use_dev_spaces(
+                cluster_name, resource_group_name,
+                space_name, parent_space_name)
```
Comment (troydai) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: What is this for? This introduces an implicit dependency from the CLI to the extension module. This is significant change we can't ignore.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py [hunk 7]
```diff
@@ -197,6 +197,17 @@ def load_arguments(self, _):
         c.argument('os_type', get_enum_type(aci_connector_os_type))
         c.argument('service_principal')
 
+    with self.argument_context('aks use-dev-spaces') as c:
+        c.argument('cluster_name', options_list=['--name', '-n'])
+        c.argument('resource_group_name', options_list=['--resource-group', '-g'])
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py]: You don't need this.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 8]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+    Using 'az configure --defaults group=<name>'.
+    :type resource_group_name: String
+    :param space_name: Name of the dev space to use.
+    :type space_name: String
+    :param parent_space_name: Name of a parent dev space to inherit from when creating a new dev space. \
+    By default, if there is already a single dev space with no parent, the new space inherits from this one.
+    :type parent_space_name: String
+    """
+
+    dev_spaces_extension = 'dev-spaces-preview'
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Make this a global constant

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 9]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+            from azure.cli.core.extension import get_extension_path
+            ext_dir = get_extension_path(dev_spaces_extension)
+            sys.path.append(ext_dir)
+            # Import the extension module
+            from importlib import import_module
+            azext_custom = import_module('azext_dev_spaces_preview.custom')
+            azext_custom.ads_use_dev_spaces(
+                cluster_name, resource_group_name,
+                space_name, parent_space_name)
+        except Exception as e:  # nopa pylint: disable=broad-except
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Are you able to catch a more specific error?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 10]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+    By default, if there is already a single dev space with no parent, the new space inherits from this one.
+    :type parent_space_name: String
+    """
+
+    dev_spaces_extension = 'dev-spaces-preview'
+
+    should_create_resource = False
+    if not _is_dev_spaces_extension_installed(dev_spaces_extension):
+        if _install_dev_spaces_extension(dev_spaces_extension):
+            should_create_resource = True
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: should_create_resource = _install_dev_spaces_extension....

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 11]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+    :param cluster_name: Name of the managed cluster.
+    :type cluster_name: String
+    :param resource_group_name: Name of resource group. You can configure the default group. \
+    Using 'az configure --defaults group=<name>'.
+    :type resource_group_name: String
+    :param prompt: Do not prompt for confirmation.
+    :type prompt: bool
+    """
+
+    dev_spaces_extension = 'dev-spaces-preview'
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: similar to above

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 12]
```diff
@@ -1418,6 +1418,102 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
...
+
+    should_delete_resource = False
+    if not _is_dev_spaces_extension_installed(dev_spaces_extension):
+        if _install_dev_spaces_extension(dev_spaces_extension):
+            should_delete_resource = True
+    else:
+        should_delete_resource = True
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: similar to above

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 3 above)
Comment (saurabsa) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: This is just a summary of the method. It is not accessible to the user.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 13]
```diff
@@ -1418,12 +1418,15 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
     return sdk_no_wait(no_wait, client.create_or_update, resource_group_name, name, instance)
 
 
-def aks_use_dev_spaces(cmd, client, cluster_name, resource_group_name, space_name='default', parent_space_name=None):  # pylint: disable=line-too-long
+dev_spaces_extension = 'dev-spaces-preview'
```
Comment (troydai) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: The constant variable name should be all upper cases.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 14]
```diff
@@ -1434,83 +1437,67 @@ def aks_use_dev_spaces(cmd, client, cluster_name, resource_group_name, space_nam
...
-            ext_dir = get_extension_path(dev_spaces_extension)
-            sys.path.append(ext_dir)
-            # Import the extension module
-            from importlib import import_module
-            azext_custom = import_module('azext_dev_spaces_preview.custom')
             azext_custom.ads_remove_dev_spaces(
-                cluster_name, resource_group_name, prompt)
-        except Exception as e:  # nopa pylint: disable=broad-except
-            raise CLIError(e)
+                name, resource_group_name, prompt)
```
Comment (troydai) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Keep it on the same line to keep the code concise.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 15]
```diff
@@ -1434,83 +1437,67 @@ def aks_use_dev_spaces(cmd, client, cluster_name, resource_group_name, space_nam
...
         from azure.cli.command_modules.extension import custom
-        custom.add_extension(extension_name=dev_spaces_extension)
+        custom.add_extension(extension_name=extension_name)
     except Exception:  # nopa pylint: disable=broad-except
         return False
     return True
 
 
-def _is_dev_spaces_extension_installed(dev_spaces_extension):
+def _is_dev_spaces_extension_installed(extension_name):
```
Comment (troydai) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Undesirable function name. The name indicates this method is a read-only getter. However, the method actually installs extension when the extension is missing. A better name will look like `get_or_add_extension`

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 16]
```diff
@@ -1434,83 +1437,67 @@ def aks_use_dev_spaces(cmd, client, cluster_name, resource_group_name, space_nam
...
-def _install_dev_spaces_extension(dev_spaces_extension):
+def _get_azext_module(extension_name):
+    try:
+        # Adding the installed extension in the path
+        from azure.cli.core.extension import get_extension_path
+        ext_dir = get_extension_path(extension_name)
+        sys.path.append(ext_dir)
+        # Import the extension module
+        from importlib import import_module
+        azext_custom = import_module('azext_dev_spaces_preview.custom')
```
Comment (troydai) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Extract this string as constant as well.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 15 above)
Comment (saurabsa) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Will go with _get_or_add_extension()

**Comment by saurabsa** [comment]
I have tested both the azure-cli and the dependent azure-cli-extension and it works fine. Will do more testing before the CI gets green to be safe.

**Pull request merged**